♻️ Union list support

prashantgupta24 opened the pull request:
Add example to support simple lists and lists in Union!

Heavy lifting done by this PR: https://github.com/IBM/py-to-proto/pull/54

fix https://github.com/caikit/caikit/issues/198

(also please squash commits, unable to git squash when there's a `merge main` commit)

prashantgupta24 commented on caikit/runtime/utils/servicer_util.py:
```diff
@@ -47,8 +47,15 @@ def validate_caikit_library_class_exists(cdm, class_name):
     except AttributeError as e:
-        log.error("<RUN24024010E>", "No Caikit Library CDM class for %s", class_name)
-        raise e
+        log.warning("<RUN24024050W>", "No Caikit Library CDM class for %s", class_name)
```
Based on internal discussions, I'm also okay in removing the whole `validate_data_model` function altogether!

joerunde replied:
Here's a story: https://github.com/caikit/caikit/issues/247

prashantgupta24 replied:
Will handle in subsequent PR

joerunde commented on tests/core/data_model/test_dataobject.py:
```diff
@@ -878,6 +878,51 @@ class Derived(Base):
+
+    # proto round trip
+    foo_int = Foo.IntSequence(values=[1, 2])
+    foo1 = Foo(foo=foo_int)
```
Is this the only way these can be constructed?

It's good, but I think a native python @dataclass user would expect to be able to write
```
foo_with_ints = Foo([1, 2, 3])
foo_with_strs = Foo(["1", "2", "3"])
```

prashantgupta24 replied:
Hmm, yeah that would be ideal, but can that be a separate feature request? This PR is mostly to handle the lists within function arguments, I believe in order to support the above, there _might_ have to be metaclass changes? Not sure

joerunde commented on tests/runtime/utils/test_servicer_util.py:
```diff
@@ -68,7 +68,7 @@ def test_servicer_util_validate_caikit_library_class_exists_returns_caikit_class
 def test_servicer_util_validate_caikit_library_class_exists_raises_for_garbage_input():
     """Test that validate_caikit_library_class_exists raises for garbage 'model'"""
-    with pytest.raises(AttributeError):
+    with pytest.raises(ValueError):
```
I don't understand how your changes made this now be a value error? 🤷 

But I'm okay with it if we're discussing removing these checks anyway

joerunde commented on tests/fixtures/sample_lib/modules/other_task/other_implementation.py:
```diff
@@ -51,11 +51,13 @@ def train(
         cls,
         training_data: DataStream[int],
         sample_input: Union[SampleInputType, str],
+        list_value: Union[List[str], List[int]],
```
I like the completeness of testing, but I do worry that we don't have much organization of these test fixtures.
There is a "primitive party implementation" test module, should we put all of these python primitive cases there instead?

prashantgupta24 replied:
Can we open a separate issue on organization of test fixtures and tackle it there?

joerunde replied:
> Actually let me change that in this PR itself

Yup, I'd rather do it this way. Thanks!

Should this PR wait to merge until the dict PR merges?

joerunde reviewed (changes requested):
I probably missed the design discussion around how to instantiate these things in python, did we decide to not support simple lists or is that just not in a test? (Or did I just miss it?)

joerunde commented:
Yeah I don't know about the git history problem, but I'm fine if we just use the squash merge button when merging if you want

prashantgupta24 commented:
> did we decide to not support simple lists or is that just not in a test? (Or did I just miss it?)

I can add a test for that, I'm assuming the only thing missing is a test for it

prashantgupta24 commented:
Hey @joerunde, update on this PR:

- added the test for a simple list
- moved test to `primitive_party`
- separate issue for removing validation of data model
- separate issue for python style initialization of lists within unions

joerunde reviewed (approved):
lpgtm!
